### PR TITLE
SM2/SM3/SM4: Chinese cipher support

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

Add support for:
 - SM2 elliptic curve and SM2 sign/verify
 - SM3 digest
 - SM4 cipher with modes ECB/CBC/CTR/GCM/CCM

Add APIs for SM3 and SM4.
Add SM2 sign and verify APIs.
Add support for SM3 in wc_Hash and wc_Hmac API.
Add support for SM3 and SM4 through EVP layer.
Add support for SM2-SM3 certificates. Support key ID and name hash being with SHA-1/256 or SM3.
Add support for TLS 1.3 cipher suites: TLS-SM4-GCM-SM3, TLS-SM4-CCM-SM3 Add support for TLS 1.2 SM cipher suite: ECDHE-ECDSA-SM4-CBC-SM3 Add support for SM3 in wc_PRF_TLS.
Add SM2-SM3 certificates and keys. Generated with GmSSL-3.0.0 and OpenSSL.

# Testing

Tested using private implementation of SM algorithms.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: wolfcrypt/src/sm3.c [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+#ifdef WOLFSSL_SM3
```
Comment (dgarske) [inline wolfcrypt/src/sm3.c:2]: Please include the wolfssl/wolfcrypt/settings.h for these empty files.

File: wolfcrypt/src/ecc.c [hunk 2]
```diff
@@ -101,6 +101,7 @@ ECC Curve Types:
  * HAVE_ECC_SECPR3      Enables SECP R3 curves                  default: off
  * HAVE_ECC_BRAINPOOL   Enables Brainpool curves                default: off
  * HAVE_ECC_KOBLITZ     Enables Koblitz curves                  default: off
+ * WOLFSSL_SM2         Enables SM2 curves                      default: off
```
Comment (dgarske) [inline wolfcrypt/src/ecc.c:104]: Alignment in comment off.

**Review by dgarske — CHANGES_REQUESTED** [review]
Wow, great work on this.

File: wolfcrypt/src/sm3.c (hunk 1 above)
Comment (SparkiDev) [reply wolfcrypt/src/sm3.c:8]: fixed

File: wolfcrypt/src/ecc.c (hunk 2 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c]: fixed

File: src/internal.c [hunk 3]
```diff
@@ -2742,14 +2742,19 @@ void FreeCiphers(WOLFSSL* ssl)
...
     XFREE(ssl->decrypt.aes, ssl->heap, DYNAMIC_TYPE_CIPHER);
 #endif
+#if defined(WOLFSSL_SM4_GCM) || defined(WOLFSSL_SM4_CCM)
+    wc_Sm4Free(ssl->encrypt.sm4);
+    wc_Sm4Free(ssl->decrypt.sm4);
+    XFREE(ssl->encrypt.sm4, ssl->heap, DYNAMIC_TYPE_CIPHER);
+    XFREE(ssl->decrypt.sm4, ssl->heap, DYNAMIC_TYPE_CIPHER);
+#endif
+#if !defined(WOLFSSL_NO_TLS12)
+    XFREE(ssl->decrypt.additional, ssl->heap, DYNAMIC_TYPE_SM4_BUFFER);
```
Comment (dgarske) [inline src/internal.c:2755]:
```Testing CONFIG_REMAINDER = --enable-dtls --enable-dtls13 --enable-dtls-mtu  --enable-opensslextra
make[3]: warning: -j3 forced in submake: resetting jobserver mode.
make[3]: warning: -j3 forced in submake: resetting jobserver mode.
make[4]: warning: -j3 forced in submake: resetting jobserver mode.
In file included from ./wolfssl/internal.h:27,
                 from src/internal.c:102:
src/internal.c: In function ‘FreeCiphers’:
src/internal.c:2755:23: error: ‘Ciphers’ has no member named ‘additional’
 2755 |     XFREE(ssl->decrypt.additional, ssl->heap, DYNAMIC_TYPE_SM4_BUFFER);
      |                       ^
./wolfssl/wolfcrypt/types.h:560:63: note: in definition of macro ‘XFREE’
  560 |                     #define XFREE(p, h, t)       {void* xp = (p); if (xp) wolfSSL_Free(xp);}
      |                                                               ^
src/internal.c:2756:23: error: ‘Ciphers’ has no member named ‘additional’
 2756 |     XFREE(ssl->encrypt.additional, ssl->heap, DYNAMIC_TYPE_SM4_BUFFER);
      |                       ^
./wolfssl/wolfcrypt/types.h:560:63: note: in definition of macro ‘XFREE’
  560 |                     #define XFREE(p, h, t)       {void* xp = (p); if (xp) wolfSSL_Free(xp);}
      |                                                               ^
make[2]: *** [Makefile:6808: src/libwolfssl_la-internal.lo] Error 1
make[2]: *** Waiting for unfinished jobs....```

File: wolfssl/wolfcrypt/types.h [hunk 4]
```diff
@@ -1088,21 +1089,28 @@ typedef struct w64wrapper {
         WC_HASH_TYPE_BLAKE2B = 14,
         WC_HASH_TYPE_BLAKE2S = 15,
         #define _WC_HASH_TYPE_MAX WC_HASH_TYPE_BLAKE2S
+        #ifdef WOLFSSL_SM3
+            WC_HASH_TYPE_SM3     = 16,
```
Comment (dgarske) [inline wolfssl/wolfcrypt/types.h:1093]: Can/should we avoid changing these numbers?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 3 above)
Comment (SparkiDev) [reply src/internal.c:2755]: fixed

File: wolfssl/wolfcrypt/types.h (hunk 4 above)
Comment (SparkiDev) [reply wolfssl/wolfcrypt/types.h]: fixed - restored numbering and added it to the end

**Comment by guidovranken** [comment]
Ping me when the PR is ready please; I'll test it.

**Comment by SparkiDev** [comment]
retest this please

**Comment by SparkiDev** [comment]
retest this please

**Comment by SparkiDev** [comment]
retest this please

**Comment by SparkiDev** [comment]
retest this please

**Review by dgarske — CHANGES_REQUESTED** [review]
CI reports (PRB-aes.txt_10):

```
wolfcrypt/src/aes.c: In function ‘GHASH_UPDATE’:
wolfcrypt/src/aes.c:5427:32: error: ‘Aes’ has no member named ‘M0’
 5427 |         GMULT(AES_TAG(aes), aes->M0);                   \
      |                                ^~
wolfcrypt/src/aes.c:6359:17: note: in expansion of macro ‘GHASH_ONE_BLOCK’
 6359 |                 GHASH_ONE_BLOCK(aes, AES_LASTGBLOCK(aes));
      |                 ^~~~~~~~~~~~~~~
wolfcrypt/src/aes.c:5427:32: error: ‘Aes’ has no member named ‘M0’
 5427 |         GMULT(AES_TAG(aes), aes->M0);                   \
      |                                ^~
wolfcrypt/src/aes.c:6373:13: note: in expansion of macro ‘GHASH_ONE_BLOCK’
 6373 |             GHASH_ONE_BLOCK(aes, a);
      |             ^~~~~~~~~~~~~~~
wolfcrypt/src/aes.c:5427:32: error: ‘Aes’ has no member named ‘M0’
 5427 |         GMULT(AES_TAG(aes), aes->M0);                   \
      |                                ^~
wolfcrypt/src/aes.c:6388:9: note: in expansion of macro ‘GHASH_ONE_BLOCK’
 6388 |         GHASH_ONE_BLOCK(aes, AES_LASTGBLOCK(aes));
      |         ^~~~~~~~~~~~~~~
wolfcrypt/src/aes.c:5427:32: error: ‘Aes’ has no member named ‘M0’
 5427 |         GMULT(AES_TAG(aes), aes->M0);                   \
      |                                ^~
wolfcrypt/src/aes.c:6408:17: note: in expansion of macro ‘GHASH_ONE_BLOCK’
 6408 |                 GHASH_ONE_BLOCK(aes, AES_LASTGBLOCK(aes));
      |                 ^~~~~~~~~~~~~~~
wolfcrypt/src/aes.c:5427:32: error: ‘Aes’ has no member named ‘M0’
 5427 |         GMULT(AES_TAG(aes), aes->M0);                   \
      |                                ^~
wolfcrypt/src/aes.c:6422:13: note: in expansion of macro ‘GHASH_ONE_BLOCK’
 6422 |             GHASH_ONE_BLOCK(aes, c);
      |             ^~~~~~~~~~~~~~~
wolfcrypt/src/aes.c: In function ‘GHASH_FINAL’:
wolfcrypt/src/aes.c:5427:32: error: ‘Aes’ has no member named ‘M0’
 5427 |         GMULT(AES_TAG(aes), aes->M0);                   \
      |                                ^~
wolfcrypt/src/aes.c:6454:9: note: in expansion of macro ‘GHASH_ONE_BLOCK’
 6454 |         GHASH_ONE_BLOCK(aes, AES_LASTGBLOCK(aes));
      |         ^~~~~~~~~~~~~~~
wolfcrypt/src/aes.c:5427:32: error: ‘Aes’ has no member named ‘M0’
 5427 |         GMULT(AES_TAG(aes), aes->M0);                   \
      |                                ^~
wolfcrypt/src/aes.c:6310:9: note: in expansion of macro ‘GHASH_ONE_BLOCK’
 6310 |         GHASH_ONE_BLOCK(aes, scratch);            \
      |         ^~~~~~~~~~~~~~~
wolfcrypt/src/aes.c:6457:5: note: in expansion of macro ‘GHASH_LEN_BLOCK’
 6457 |     GHASH_LEN_BLOCK(aes);
      |     ^~~~~~~~~~~~~~~
make[2]: *** [Makefile:6122: wolfcrypt/src/src_libwolfssl_la-aes.lo] Error 1
```

**Comment by SparkiDev** [comment]
retest this please

**Review by dgarske — APPROVED** [review]

**Pull request merged**